README.md
# minterjs-util

[![NPM Package](https://img.shields.io/npm/v/minterjs-util.svg?style=flat-square)](https://www.npmjs.org/package/minterjs-util)
[![Build Status](https://img.shields.io/github/workflow/status/MinterTeam/minterjs-util/Test?label=test&style=flat-square)](https://github.com/MinterTeam/minterjs-util/actions/workflows/test-with-coverage.yml)
[![Coverage Status](https://img.shields.io/coveralls/github/MinterTeam/minterjs-util/master.svg?style=flat-square)](https://coveralls.io/github/MinterTeam/minterjs-util?branch=master)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg?style=flat-square&label=license)](https://github.com/MinterTeam/minterjs-util/blob/master/LICENSE)

A collection of utility functions for Minter

Please note that this package is under active development and is subject to change.

It is complemented by the following packages:
- [minter-js-sdk](https://github.com/MinterTeam/minter-js-sdk) complete JS solution to work with Minter
- [minterjs-wallet](https://github.com/MinterTeam/minterjs-wallet) take care of BIP39 mnemonic phrase, private key and address

## Install

`npm install minterjs-util`

or from browser

```html
<script src="https://unpkg.com/minterjs-util"></script>
<script>
const pips = minterUtil.convertFromPip(1);
const fee = (new minterUtil.FeePrice({/* ... */})).getFeeValue('0x01');
</script>
```

### FeePrice.getFeeValue()
Params:
- txType: number or string, tx type
- options: object
- options.payload: string or Buffer, tx payload
- options.payloadLength: number, length of payload
- options.coinSymbol: string, coin symbol if tx is coin creation (can be replaced with `coinSymbolLength`)
- options.coinSymbolLength: number, coin symbol length if tx is coin creation (can be replaced with `coinSymbol`)
- options.multisendCount: number, count of recipients if tx is multisend

Full example: [github.com/MinterTeam/minterjs-util/blob/master/test/fee.test.js](https://github.com/MinterTeam/minterjs-util/blob/master/test/fee.test.js)
```
import { FeePrice, TX_TYPE } from 'minterjs-util';
const getFeeValue = (new FeePrice({/* ... */})).getFeeValue;

getFeeValue(TX_TYPE.SEND);
// 0.01

getFeeValue(TX_TYPE.SEND, {payload: 'asé'});
// 0.018

getFeeValue(TX_TYPE.CREATE_COIN, {coinSymbol: 'ABCDEFG'});
// 10000

getFeeValue(TX_TYPE.CREATE_COIN, {coinSymbolLength: 6});
// 100000

getFeeValue(TX_TYPE.MULTISEND, {deltaItemCount: 5});
// 0.035
```

## License

MIT License


CHANGELOG.md
## 0.25.1 - 2023.04.05
- add estimation limit calculation methods

## 0.25.0 - 2023.04.05
- add wrapped Big.js
- update deps
- **BREAKING** maybe breaking if used cjs version with node version not supported es modules

## 0.24.2 - 2022.11.17
- add new variables

## 0.24.0 - 2022.06.02
- **BREAKING** change `FeePrice#getFeeValue` to throw by default on invalid coinSymbol length for coin creation tx
- add `options.fallbackOnInvalidInput` field to `FeePrice#getFeeValue` to fallback if invalid coinSymbol or deltaItemCount

## 0.23.0 - 2022.04.12
- add `lock` and `lock_stake` tx types
- add `isDisabled` property to `txTypeList`
- **BREAKING** change `isMinterPrefixed` to accept prefix without value
- **BREAKING** change `coinToBuffer' to accept undefined and treat it as empty string

## 0.22.0 - 2021.12.01
- add `add_limit_order`, `remove_limit_order` tx types

## 0.21.0 - 2021.04.09
- add v2.0 tx types
- add `payloadLength` param to `getFeeValue`
- **BREAKING** fee module now designed as constructor, which accepts current network fee values and returns instance with method `getFeeValue`
- **BREAKING** TX_TYPE.EDIT_COIN_OWNER renamed to TX_TYPE.EDIT_TICKER_OWNER
- **BREAKING** remove deprecated `numToBig` and default export from converter module, use `numberToBig` and named exports instead
- **BREAKING** converter and coin-math now rounds to half even to align with blockchain rounding

## 0.20.0 - 2020.12.30
- **BREAKING** change fee calculation for Create Coin tx (now it depends on unit value)
- add `coinSymbolLength` param to `getFeeValue`

## 0.19.0 - 2020.12.22
- **BREAKING** change unit value, which is used in fee calculation, from 0.001 to 0.1

## 0.18.1
- deprecate `numToBig`, use `numberToBig` instead

## 0.18.0
- add chilinet tx types and fees

## 0.17.0 - 2020-08-7
- update deps

## 0.16.2 - 2020-05-31
- export variables

## 0.16.1 - 2020-05-31
- move here tx-types and coin helpers from [minterjs-tx](https://github.com/MinterTeam/minterjs-tx)

## 0.15.2 - 2020-05-30
- export `numToBig` function

## 0.15.1 - 2020-01-29
- fix naming typo

## 0.15.0 - 2019-12-16
- add `addressToString` method
- add `checkToString` method
- allow `publicToString` to accept `0x` prefixed strings

## 0.14.1 - 2019-11-13
- update `toBuffer` error message on invalid string

## 0.14.0 - 2019-11-12
- **BREAKING** toBuffer accepts only Minter prefixed and `0x` prefixed string params now, you have to use `Buffer.from(str, 'utf-8')` if you want pass arbitrary string
- drop safe-buffer dependency

## 0.13.1 - 2019-10-24
- lighten converter.js by using part of previous deps

## 0.13.0 - 2019-10-01
- **BREAKING** throws on invalid `multisendCount` param
- fix multisend fee calculation

## 0.12.0 - 2019-08-15
- **BREAKING** coin math functions now return string instead of number

## 0.11.0 - 2019-08-14
- **BREAKING** update `getFeeValue`, `payloadLength` param changed to `payload` string and moved to options object, `coinSymbolLength` param changed to `coinSymbol` string
- **BREAKING** update `getCoinSymbolFee`, `coinSymbolLength` param changed to `coinSymbol` string

## 0.10.1 - 2019-07-30
- fix browser usage of UMD module
- tweak bundle tests

# 0.10.0 - 2019-07-10
- add coin math functions

# 0.9.1 - 2019-07-08
- update deps

# 0.9.0 - 2019-06-13
- Added: `isValidTransaction` methods

# 0.8.1 - 2019-06-05
- update deps

# 0.8.0 - 2019-05-14
- **BREAKING** Removed: moved `defineProperties` to [minterjs-tx](https://github.com/MinterTeam/minterjs-tx) package

# 0.7.3 - 2019-05-12
- Fixed: update CreateCoin fee to return minimal fee when `coinSymbolLength` is not specified

# 0.7.2 - 2019-05-12
- Fixed: update CreateCoin tx commission according to minter-node [v0.19.0](https://github.com/MinterTeam/minter-go-node/blob/master/CHANGELOG.md#0190)

# 0.7.1 - 2019-04-10
- Changed: `getFeeValue` returns `false` instead of throwing

# 0.7.0 - 2019-04-09
- Added: `getFeeValue` method

# 0.6.1 - 2019-02-28
- fix package.json "browser" field

# 0.6.0 - 2019-02-27
- add UMD and commonjs builds

# 0.5.0 - 2019-02-20
- **BREAKING** converting to hex now pad to even (`11` will become `0b` instead of `b`)

# 0.4.0 - 2019-02-19
- **BREAKING** converting from pip to bip in hex format will throw an error
- **BREAKING** converting invalid number will throw instead of returning `NaN`
- replace `bignumber.js` dependency with `big.js` to reduce size

# 0.3.3 - 2019-02-18
- update deps
- fix ethereumjs-util

# 0.3.2 - 2019-02-18
- broken version

# 0.3.1 - 2019-02-06
- add `nonBinaryArrayTransform` option to `defineProperties()`

# 0.3.0 - 2019-02-06
- add `defineProperties()` method
- update deps

# 0.2.3 - 2018-12-06
- add `privateToAddressString()` methods

## 0.2.2 - 2018-11-30
- fix converter to not produce exponential values

## 0.2.1 - 2018-11-29
- add option to convert to hex value

## 0.2.0 - 2018-11-26
- **BREAKING** update converter to return `string` instead of `BigNumber`

## 0.1.0 - 2018-11-24
- add pip converter utils: `convert`, `convertToPip`, `convertFromPip`. (Moved from `minterjs-tx` package)


src/tx-types.js
import {padToEven} from 'ethjs-util';

/**
 * @enum {string} TX_TYPE
 */
export const TX_TYPE = {
    SEND: '0x01',
    SELL: '0x02',
    SELL_ALL: '0x03',
    BUY: '0x04',
    CREATE_COIN: '0x05',
    DECLARE_CANDIDACY: '0x06',
    DELEGATE: '0x07',
    UNBOND: '0x08',
    REDEEM_CHECK: '0x09',
    SET_CANDIDATE_ON: '0x0A',
    SET_CANDIDATE_OFF: '0x0B',
    CREATE_MULTISIG: '0x0C',
    MULTISEND: '0x0D',
    EDIT_CANDIDATE: '0x0E',
    SET_HALT_BLOCK: '0x0F',
    RECREATE_COIN: '0x10',
    EDIT_TICKER_OWNER: '0x11',
    EDIT_MULTISIG: '0x12',
    PRICE_VOTE: '0x13',
    EDIT_CANDIDATE_PUBLIC_KEY: '0x14',
    ADD_LIQUIDITY: '0x15',
    REMOVE_LIQUIDITY: '0x16',
    SELL_SWAP_POOL: '0x17',
    BUY_SWAP_POOL: '0x18',
    SELL_ALL_SWAP_POOL: '0x19',
    EDIT_CANDIDATE_COMMISSION: '0x1A',
    MOVE_STAKE: '0x1B',
    MINT_TOKEN: '0x1C',
    BURN_TOKEN: '0x1D',
    CREATE_TOKEN: '0x1E',
    RECREATE_TOKEN: '0x1F',
    VOTE_COMMISSION: '0x20',
    VOTE_UPDATE: '0x21',
    CREATE_SWAP_POOL: '0x22',
    ADD_LIMIT_ORDER: '0x23',
    REMOVE_LIMIT_ORDER: '0x24',
    LOCK_STAKE: '0x25',
    LOCK: '0x26',
};

const DISABLED_TYPES = {
    [TX_TYPE.PRICE_VOTE]: true,
};

// swap: keys <=> values
const txTypeKeys = Object.fromEntries(Object.entries(TX_TYPE).map(([key, hexValue]) => [hexValue, key]));

/**
 * @typedef {object} TxTypeItem
 * @property {TX_TYPE} hex
 * @property {string} name
 * @property {number} number
 * @property {string} key
 * @property {boolean} isDisabled
 */

/** @type {Array<TxTypeItem>} */
const txTypeList = [];

/**
 * @param {string} hex
 * @param {string} [name]
 * @return {TxTypeItem}
 */
function fillList(hex, name) {
    const result = {};

    result.key = txTypeKeys[hex].toLowerCase();
    result.name = name || result.key.replace(/_/g, ' ');
    result.number = Number(hex);
    result.hex = hex;
    result.isDisabled = !!DISABLED_TYPES[hex];

    txTypeList[result.number] = result;

    return result;
}
fillList(TX_TYPE.SEND);
fillList(TX_TYPE.SELL);
fillList(TX_TYPE.SELL_ALL);
fillList(TX_TYPE.BUY);
fillList(TX_TYPE.CREATE_COIN);
fillList(TX_TYPE.DECLARE_CANDIDACY);
fillList(TX_TYPE.DELEGATE);
fillList(TX_TYPE.UNBOND);
fillList(TX_TYPE.REDEEM_CHECK);
fillList(TX_TYPE.SET_CANDIDATE_ON);
fillList(TX_TYPE.SET_CANDIDATE_OFF);
fillList(TX_TYPE.CREATE_MULTISIG);
fillList(TX_TYPE.MULTISEND);
fillList(TX_TYPE.EDIT_CANDIDATE);
fillList(TX_TYPE.SET_HALT_BLOCK, 'vote for halt block');
fillList(TX_TYPE.RECREATE_COIN);
fillList(TX_TYPE.EDIT_TICKER_OWNER);
fillList(TX_TYPE.EDIT_MULTISIG);
fillList(TX_TYPE.PRICE_VOTE, 'vote for price');
fillList(TX_TYPE.EDIT_CANDIDATE_PUBLIC_KEY);
fillList(TX_TYPE.ADD_LIQUIDITY, 'add liquidity to pool');
fillList(TX_TYPE.REMOVE_LIQUIDITY, 'remove liquidity from pool');
fillList(TX_TYPE.SELL_SWAP_POOL, 'sell within pool');
fillList(TX_TYPE.BUY_SWAP_POOL, 'buy within pool');
fillList(TX_TYPE.SELL_ALL_SWAP_POOL, 'sell all within pool');
fillList(TX_TYPE.EDIT_CANDIDATE_COMMISSION);
fillList(TX_TYPE.MOVE_STAKE);
fillList(TX_TYPE.MINT_TOKEN);
fillList(TX_TYPE.BURN_TOKEN);
fillList(TX_TYPE.CREATE_TOKEN);
fillList(TX_TYPE.RECREATE_TOKEN);
fillList(TX_TYPE.VOTE_COMMISSION, 'vote for commission price');
fillList(TX_TYPE.VOTE_UPDATE, 'vote for network update');
fillList(TX_TYPE.CREATE_SWAP_POOL);
fillList(TX_TYPE.ADD_LIMIT_ORDER);
fillList(TX_TYPE.REMOVE_LIMIT_ORDER);
fillList(TX_TYPE.LOCK_STAKE);
fillList(TX_TYPE.LOCK);

export {txTypeList};

/**
 *
 * @param {TX_TYPE|number|string|Buffer|Uint8Array|Array<number>} txType
 * @return {TX_TYPE}
 */
export function normalizeTxType(txType) {
    function throwInvalid() {
        throw new Error(`Invalid tx type: ${txType}`);
    }

    if (!txType) {
        throwInvalid();
    }
    // Buffer or Uint8Array to TX_TYPE
    if (txType instanceof Uint8Array || Array.isArray(txType)) {
        if (txType.length === 0) {
            throwInvalid();
        }
        txType = Buffer.from(txType).toString('hex');
        txType = `0x${txType}`;
    }
    // invalid string to number
    if (typeof txType === 'string' && txType.indexOf('0x') !== 0 && txType.indexOf('0X') !== 0) {
        txType = Number.parseInt(txType, 10);
    }
    // number to TX_TYPE
    if (typeof txType === 'number') {
        txType = padToEven(txType.toString(16));
        txType = `0x${txType}`;
    }

    txType = txType.toUpperCase();
    txType = txType.replace(/^0X/, '0x');

    if (!Object.values(TX_TYPE).includes(txType)) {
        throw new Error('Invalid tx type');
    }

    return txType;
}
